src/Compilers/CSharp/Portable/Binder/PatternExplainer.cs

```diff
@@ -0,0 +1,512 @@
+                var properties = new Dictionary<Symbol, string>();
+                bool needsPropertyString = false;
+
+                foreach (var eval in evaluations)
+                {
+                    switch (eval)
+                    {
+                        case BoundDagDeconstructEvaluation e:
+                            var method = e.DeconstructMethod;
+                            int extensionExtra = method.RequiresInstanceReceiver ? 0 : 1;
+                            int count = method.Parameters.Length - extensionExtra;
+                            var subpatterns = new ArrayBuilder<string>(count);
+                            for (int j = 0; j < count; j++)
+                            {
+                                var elementTemp = new BoundDagTemp(e.Syntax, method.Parameters[j + extensionExtra].Type, e, j);
+                                var newPattern = SamplePatternForTemp(elementTemp, constraintMap, evaluationMap, requireExactType: false);
+                                subpatterns.Add(newPattern);
+                            }
+                            var result = "(" + string.Join(", ", subpatterns) + ")";
+                            if (deconstruction != null && needsPropertyString)
+                            {
+                                deconstruction = deconstruction + " { }";
+                                needsPropertyString = properties.Count != 0;
+                            }
+
+                            deconstruction = (deconstruction is null) ? result : deconstruction + " and " + result;
+                            needsPropertyString |= count == 1;
+                            break;
+                        case BoundDagFieldEvaluation e:
+                            {
+                                var subInput = new BoundDagTemp(e.Syntax, e.Field.Type, e);
+                                var subPattern = SamplePatternForTemp(subInput, constraintMap, evaluationMap, false);
+                                properties.Add(e.Field, subPattern);
+                            }
+                            break;
+                        case BoundDagPropertyEvaluation e:
+                            {
+                                var subInput = new BoundDagTemp(e.Syntax, e.Property.Type, e);
+                                var subPattern = SamplePatternForTemp(subInput, constraintMap, evaluationMap, false);
+                                properties.Add(e.Property, subPattern);
+                            }
+                            break;
+                        default:
+                            return null;
+                    }
+                }
+
+                needsPropertyString |= deconstruction == null && typeName == null || properties.Count != 0;
+                var propertyString = needsPropertyString ? (deconstruction != null ? " {" : "{") + string.Join(", ", properties.Select(kvp => $" {kvp.Key.Name}: {kvp.Value}")) + " }" : null;
```

Comment:
>properties.Select [](start = 117, length = 17)

Can we use the order from evaluations instead (is that order meaningful), or sort the properties, rather than using an arbitrary order?

Reply:
This does exactly that. A dictionary from which nothing has been removed iterates the values in the order in which they were inserted.

---
In reply to: [443079309](https://github.com/dotnet/roslyn/pull/44702#discussion_r443079309) [](ancestors = 443079309)